serf/serf.go

```diff
@@ -593,9 +595,9 @@ func (s *Serf) Join(existing []string, ignoreOld bool) (int, error) {
 	// Ignore any events from a potential join. This is safe since we hold
 	// the joinLock and nobody else can be doing a Join
 	if ignoreOld {
-		s.eventJoinIgnore = true
+		s.eventJoinIgnore.Store(true)
```

Comment:
The variable is accessed once with and once without the lock in different methods which in consul is called from different go routines. This is the data race since this write isn't sync'ed with the read in the other function. You also can't use the same lock to guard both places since they seem to depend on each other and sort-of deadlock.